File: java/01140_stone_game_II.java
import java.util.HashMap;
import java.util.Map;

class Solution {

    public int stoneGameII(int[] piles) {
        return getMaxScore(true, 1, 1, 0, 0, piles, new HashMap<>());
    }

    private int getMaxScore(boolean turn, int m, int x, int index, int maxScore, int[] piles, Map<String, Integer> memo) {
        if (memo.containsKey(index + "_" + x)) {
            return memo.get(index + "_" + x);
        }
        m = Math.max(m, x);
        int tempScore = 0;
        if (turn) {
            for (int i = index; i < index + 2 * m && i < piles.length; i++) {
                tempScore += piles[i];
                int currentScore = tempScore + getMaxScore(false, m, i - index + 1, index + 1, maxScore, piles, memo);
                maxScore = Math.max(maxScore, currentScore);
            }
        } else {
            for (int i = index; i < index + 2 * m && i < piles.length; i++) {
                int currentScore = getMaxScore(false, m, i - index + 1, index + 1, maxScore, piles, memo);
                maxScore = Math.max(maxScore, currentScore);
            }
        }
        memo.put(index + "_" + x, maxScore);
        return maxScore;
    }
}

File: python/01140_stone_game_II.py
#     Python
# Time complexity - O(n^3)
# Space complexity - O(n^2)
class Solution:
    def stoneGameII(self, piles: List[int]) -> int:
        def getMaxScore(turn: bool, m: int, index: int, memo: dict) -> int:
            if index >= len(piles):
                return 0

            key = (index, m, turn)
            if key in memo:
                return memo[key]

            maxScore = float('-inf') if turn else float('inf')
            tempScore = 0

            for x in range(1, 2 * m + 1):
                if index + x > len(piles):
                    break

                tempScore += piles[index + x - 1]

                if turn:
                    maxScore = max(maxScore, tempScore + getMaxScore(False, max(m, x), index + x, memo))
                else:
                    maxScore = min(maxScore, getMaxScore(True, max(m, x), index + x, memo))

            memo[key] = maxScore
            return maxScore

        memo = {}
        return getMaxScore(True, 1, 0, memo)
